### ISSUE TITLE
Resource nesting

### ISSUE THREAD

**Issue by openmosix** [Gild/gild-api#10]
I'll start this - given there was an internal debate on resource representation

https://github.com/Gild/gild-api#minimize-path-nesting

Any thoughts about this?

cc: @andreacfm @taganaka @tejo

**Comment by openmosix** [comment]
cc: @liquid1982 @stefanoc

**Comment by stefanoc** [comment]
I agree that nesting should be limited to scoped relationships (has-many or has-one).

**Comment by andreacfm** [comment]
:+1:  About limiting the path nesting to one level max.
What about nesting path when reading list of resources:

```ruby 
GET /book/1/authors

vs

GET /authors?book_id=1
```

The possible issue is having many endpoints that returns list of authors in place of having one endpoint filterable by params.

Thoughts?

**Comment by taganaka** [comment]
I agree with @stefanoc 

Unpacking something like
```
/companies/{company_id}/jobs/{job_id}/profiles/{profile_id}
```

To: 

```
/companies/{company_id}
/companies/{company_id}/jobs
/jobs/{job_id}
/jobs/{job_id}/profiles
/profiles/{profile_id}
```

Is less ambiguous and it is very clear on how a method should be implemented

**Comment by andreacfm** [comment]
@taganaka 

A question. What do you think about this version?

```ruby
/companies/{company_id}
/jobs?company_id={company_id}
/jobs/{job_id}
/profiles?job_id={job_id}
/profiles/{profile_id}
```

**Comment by liquid1982** [comment]
I think that one important aspect to take in account is the access level of the user who is consuming the API.

For example, we have 2 main user categories (platform users and candidates).

For platform users, we want to filter resources based on a particular scope (their company), so we should use something like:

```
/companies/{company_id}/jobs
```

For candidates, we want to allow read access to the same resource (jobs), but we also want to apply some sort of filters (company, location, whatever). So the endpoint would be

```
/jobs?filter[company][domain]=example.com&filter[location]=San+Francisco
```

Using `filter` as key makes the action explicit and is compliant with jsonapi.

**Comment by andreacfm** [comment]
After thinking deeply about this I arrived to the conclusion that endpoints duplication based on user role/state could become an issue. If we start to create endpoint dedicated to a particular user/role/group/state we will probably finish to maintain several endpoints that returns the same kind of resources and more or less accepts the same filters.
Api does only know and cares about the user that is calling and must be concerned to give back only what the user is authorized to get back.
If we think in these terms what we need is to be sure to apply the correct ACL rules. An example:

```ruby
# GET /articles
------------------------

get '/articles' do 
     articles = Articles.accessible_by(current_user).here_add_any_filter_you_want
end

# accessible_by is a scope that apply one or more query clauses or other logic 
# to reduce the set of resources returned.
# As as scope is chinable by default:
 
 Articles.accessible_by(current_user).where(owner: owner)
 Articles.accessible_by(current_user).find_by(title: title)
```

A defined user calling the API in a certain moment has the right to access to a defined set of resources. This does not change by endpoint. If a user can see just public articles **accessible_by** will just return public articles and applying more filters can just reduce the resources returned to a smaller group.
If you think in these terms having more endpoints that returns collections of articles, maybe scoped in the url, does not really make a lot of sense.
Is a client choice to filter the articles to fit a specific case. For example, I could create a list of all the articles a user can read including private and public and maybe decide to create a list of the articles that belongs to a company just passing an extra param on the same endpoint.
At the end of the day is just a matter of filtering what a user has right to see. A valuable side effect is that ACL scopes are not usable only in reading collections but should be widely used even when reading a single resource (GOOD: Articles.accessible_by(current_user).find(id)) 
Endpoints like:

```ruby
GET /companies/ID/articles
```
creates a sort of autofiltering that does not really add a new functionality but reduce the possibility of the client to use all the available data in a flexible way.
At the end we will need to implement both the endpoints adding complexity that could probably be avoided.

**Comment by openmosix** [comment]
Did we settle on this?
